`CartesianCentroidIT.testMultiValuedField` fails on `8.7`

### CI Link

https://gradle-enterprise.elastic.co/s/axidmlforgvci

### Repro line

./gradlew ':x-pack:plugin:spatial:internalClusterTest' --tests "org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianCentroidIT.testMultiValuedField" \   -Dtests.seed=8668813C5163E0C8 \   -Dtests.locale=mt-MT \   -Dtests.timezone=Europe/Chisinau \   -Druntime.java=20

### Does it reproduce?

Yes

### Applicable branches

8.7

### Failure history

_No response_

### Failure excerpt

```
11:46:03 org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianCentroidIT > testMultiValuedField FAILED
11:46:03     java.lang.AssertionError: Mismatching value for 'y' field of centroid
11:46:03     Expected: a numeric value within <2.0313534124233142E27> of <2.031197896205879E32>
11:46:03          but: <2.031353412423314E32> differed by <1.3520268331099276E28> more than delta <2.0313534124233142E27>
11:46:03         at __randomizedtesting.SeedInfo.seed([8668813C5163E0C8:C2B7DE705B53E615]:0)
11:46:03         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
11:46:03         at org.junit.Assert.assertThat(Assert.java:956)
11:46:03         at org.elasticsearch.search.aggregations.metrics.CentroidAggregationTestBase.assertSameCentroid(CentroidAggregationTestBase.java:160)
11:46:03         at org.elasticsearch.search.aggregations.metrics.CentroidAggregationTestBase.testMultiValuedField(CentroidAggregationTestBase.java:128)
11:46:03         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
11:46:03         at java.base/java.lang.reflect.Method.invoke(Method.java:578)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
11:46:03         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
11:46:03         at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
11:46:03         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
11:46:03         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
11:46:03         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
11:46:03         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
11:46:03         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
11:46:03         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
11:46:03         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
11:46:03         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
11:46:03         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
11:46:03         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
11:46:03         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
11:46:03         at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
11:46:03         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
11:46:03         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
11:46:03         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
11:46:03         at java.base/java.lang.Thread.run(Thread.java:1623)
```

--- a/test/framework/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractGeoTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractGeoTestCase.java
@@ -113,27 +113,25 @@ public abstract class AbstractGeoTestCase extends ESIntegTestCase {
         }
 
         List<IndexRequestBuilder> builders = new ArrayList<>();
-
-        SpatialPoint singleVal;
-        final SpatialPoint[] multiVal = new SpatialPoint[2];
-        double newMVLat, newMVLon;
+        final SpatialPoint[] allSingleVal = new SpatialPoint[numDocs];
+        final SpatialPoint[] allMultiVal = new SpatialPoint[2 * numDocs];
         for (int i = 0; i < numDocs; i++) {
-            singleVal = singleValues[i % numUniqueGeoPoints];
-            multiVal[0] = multiValues[i % numUniqueGeoPoints];
-            multiVal[1] = multiValues[(i + 1) % numUniqueGeoPoints];
+            allSingleVal[i] = singleValues[i % numUniqueGeoPoints];
+            allMultiVal[2 * i] = multiValues[i % numUniqueGeoPoints];
+            allMultiVal[2 * i + 1] = multiValues[(i + 1) % numUniqueGeoPoints];
             builders.add(
                 client().prepareIndex(IDX_NAME)
                     .setSource(
                         jsonBuilder().startObject()
-                            .array(SINGLE_VALUED_FIELD_NAME, singleVal.getX(), singleVal.getY())
+                            .array(SINGLE_VALUED_FIELD_NAME, allSingleVal[i].getX(), allSingleVal[i].getY())
                             .startArray(MULTI_VALUED_FIELD_NAME)
                             .startArray()
-                            .value(multiVal[0].getX())
-                            .value(multiVal[0].getY())
+                            .value(allMultiVal[2 * i].getX())
+                            .value(allMultiVal[2 * i].getY())
                             .endArray()
                             .startArray()
-                            .value(multiVal[1].getX())
-                            .value(multiVal[1].getY())
+                            .value(allMultiVal[2 * i + 1].getX())
+                            .value(allMultiVal[2 * i + 1].getY())
                             .endArray()
                             .endArray()
                             .field(NUMBER_FIELD_NAME, i)
@@ -141,20 +139,9 @@ public abstract class AbstractGeoTestCase extends ESIntegTestCase {
                             .endObject()
                     )
             );
-            singleCentroid = reset(
-                singleCentroid,
-                singleCentroid.getX() + (singleVal.getX() - singleCentroid.getX()) / (i + 1),
-                singleCentroid.getY() + (singleVal.getY() - singleCentroid.getY()) / (i + 1)
-            );
-            newMVLat = (multiVal[0].getY() + multiVal[1].getY()) / 2d;
-            newMVLon = (multiVal[0].getX() + multiVal[1].getX()) / 2d;
-            multiCentroid = reset(
-                multiCentroid,
-                multiCentroid.getX() + (newMVLon - multiCentroid.getX()) / (i + 1),
-                multiCentroid.getY() + (newMVLat - multiCentroid.getY()) / (i + 1)
-            );
         }
-
+        singleCentroid = computeCentroid(allSingleVal);
+        multiCentroid = computeCentroid(allMultiVal);
         assertAcked(prepareCreate(EMPTY_IDX_NAME).setMapping(SINGLE_VALUED_FIELD_NAME, "type=" + fieldTypeName()));
 
         assertAcked(
@@ -204,7 +191,7 @@ public abstract class AbstractGeoTestCase extends ESIntegTestCase {
         );
 
         for (int i = 0; i < 2000; i++) {
-            singleVal = singleValues[i % numUniqueGeoPoints];
+            SpatialPoint singleVal = singleValues[i % numUniqueGeoPoints];
             builders.add(
                 client().prepareIndex(HIGH_CARD_IDX_NAME)
                     .setSource(
@@ -261,6 +248,16 @@ public abstract class AbstractGeoTestCase extends ESIntegTestCase {
             assertThat("Hit " + i + " has wrong value", value.intValue(), equalTo(i));
         }
         assertThat(totalHits, equalTo(2000L));
+    }
+
+    private SpatialPoint computeCentroid(SpatialPoint[] points) {
+        final CompensatedSum compensatedSumX = new CompensatedSum(0, 0);
+        final CompensatedSum compensatedSumY = new CompensatedSum(0, 0);
+        for (SpatialPoint spatialPoint : points) {
+            compensatedSumX.add(spatialPoint.getX());
+            compensatedSumY.add(spatialPoint.getY());
+        }
+        return makePoint(compensatedSumX.value() / points.length, compensatedSumY.value() / points.length);
     }
 
     private void updateGeohashBucketsCentroid(final SpatialPoint location) {

--- a/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianCentroidIT.java
+++ b/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianCentroidIT.java
@@ -51,7 +51,7 @@ public class CartesianCentroidIT extends CentroidAggregationTestBase {
     @Override
     protected CartesianPoint randomPoint() {
         Point point = ShapeTestUtils.randomPointNotExtreme(false);
-        return makePoint(point.getX(), point.getY());
+        return makePoint((float) point.getX(), (float) point.getY());
     }
 
     @Override

